workers/unity/Packages/io.improbable.worker.sdk/ParameterConversion.cs

```diff
@@ -0,0 +1,233 @@
+                    ConvertEvent(eventItem, internalEvent =>
+                    {
+                        newItem->ItemUnion.Event = internalEvent;
+                        callback(newItem);
+                    });
+
+                    break;
+                default:
+                    throw new NotSupportedException("Invalid Item Type provided.");
+            }
+        }
+
+        public static unsafe void ConvertEventTracerParameters(EventTracerParameters[] parameters, EventTracerParametersCallback callback)
+        {
+            var handles = new List<WrappedGcHandle>(parameters.Length);
+            ConvertTracerParameters(handles, parameters, out var internalParameters);
+
+            fixed (CEventTrace.EventTracerParameters* parameterBuffer = internalParameters)
+            {
+                callback(parameterBuffer, handles);
+            }
+        }
+
+        private static unsafe WrappedGcHandle ConvertTracerParameter(EventTracerParameters parameter, ref CEventTrace.EventTracerParameters internalParameters)
+        {
+            WrappedGcHandle wrappedParameterObject = new WrappedGcHandle(parameter);
+
+            internalParameters.UserData = (void*) wrappedParameterObject.Get();
+            internalParameters.TraceCallback = parameter.TraceCallback == null
+                ? IntPtr.Zero
+                : Marshal.GetFunctionPointerForDelegate(CallbackThunkDelegates.TraceCallbackThunkDelegate);
+
+            return wrappedParameterObject;
+        }
+
+        public static void FreeTracerParameterHandles(IEnumerable<WrappedGcHandle> tracerParameterHandles)
+        {
+            foreach (var tracerParameterHandle in tracerParameterHandles)
+            {
+                tracerParameterHandle.Dispose();
+            }
+        }
+
+        private static void ConvertTracerParameters(List<WrappedGcHandle> handles, EventTracerParameters[] parameters,
+            out CEventTrace.EventTracerParameters[] internalParameters)
+        {
+            internalParameters = new CEventTrace.EventTracerParameters[parameters.Length];
+            for (var i = 0; i < parameters.Length; i++)
+            {
+                var parameterHandle = ConvertTracerParameter(parameters[i], ref internalParameters[i]);
+                if (parameterHandle != null)
+                {
+                    handles.Add(parameterHandle);
+                }
+            }
+        }
+
+        private unsafe class CallbackThunkDelegates
+        {
+            public static readonly CEventTrace.TraceCallback TraceCallbackThunkDelegate = TraceCallbackThunk;
+
+            [MonoPInvokeCallback(typeof(CEventTrace.TraceCallback))]
+            private static void TraceCallbackThunk(void* callbackHandlePtr, CEventTrace.Item* item)
+            {
+                var callbackHandle = GCHandle.FromIntPtr((IntPtr) callbackHandlePtr);
+                var callback = (Action<Item>) callbackHandle.Target;
```

Comment:
Thats a scary cast, is that correct?

Reply:
That's the bit that isn't working at the moment, so I don't think the cast is correct

Reply:
Discussed offline, now fixed 😄